File: apps/marketplace-api/app/models/consent.py
"""
GDPR Consent-related Pydantic models
"""
from pydantic import BaseModel, Field
from typing import Optional, Literal
from datetime import datetime


# ============================================
# COOKIE CONSENT
# ============================================

class CookieConsentRequest(BaseModel):
    """Cookie consent request model"""
    visitor_id: str = Field(..., description="Unique identifier for the visitor")
    necessary: bool = Field(default=True, description="Essential cookies (always true)")
    functional: bool = Field(default=False, description="Functional cookies")
    analytics: bool = Field(default=False, description="Analytics cookies")
    marketing: bool = Field(default=False, description="Marketing cookies")


class CookieConsentResponse(BaseModel):
    """Cookie consent response model"""
    id: str
    visitor_id: str
    user_id: Optional[str] = None
    necessary: bool
    functional: bool
    analytics: bool
    marketing: bool
    created_at: datetime
    updated_at: datetime


# ============================================
# USER CONSENT STATUS
# ============================================

class ConsentStatusResponse(BaseModel):
    """User consent status response model"""
    terms_accepted: bool
    terms_accepted_at: Optional[datetime] = None
    terms_version: Optional[str] = None
    privacy_accepted: bool
    privacy_accepted_at: Optional[datetime] = None
    privacy_version: Optional[str] = None
    marketing_consent: bool
    marketing_consent_at: Optional[datetime] = None


class UpdateMarketingConsentRequest(BaseModel):
    """Update marketing consent request model"""
    marketing_consent: bool = Field(..., description="Whether user consents to marketing")


class UpdateMarketingConsentResponse(BaseModel):
    """Update marketing consent response model"""
    marketing_consent: bool
    marketing_consent_at: datetime
    message: str


# ============================================
# CONSENT HISTORY
# ============================================

class ConsentHistoryItem(BaseModel):
    """Individual consent history record"""
    id: str
    consent_type: str
    consent_given: bool
    version: Optional[str] = None
    created_at: datetime


class ConsentHistoryResponse(BaseModel):
    """Consent history response model"""
    history: list[ConsentHistoryItem]
    total: int


# ============================================
# GDPR DATA REQUESTS
# ============================================

class GdprExportRequestResponse(BaseModel):
    """GDPR data export request response"""
    id: str
    status: str
    requested_at: datetime
    expires_at: Optional[datetime] = None
    message: str


class GdprExportDownloadResponse(BaseModel):
    """GDPR data export download response"""
    download_url: str
    expires_at: datetime
    message: str


class GdprDeletionRequestResponse(BaseModel):
    """GDPR deletion request response"""
    id: str
    status: str
    requested_at: datetime
    scheduled_deletion_at: datetime
    message: str


class GdprDeletionCancelResponse(BaseModel):
    """GDPR deletion cancel response"""
    message: str
    cancelled: bool


class GdprRequestStatusResponse(BaseModel):
    """GDPR request status response"""
    id: str
    request_type: Literal["export", "deletion"]
    status: Literal["pending", "processing", "completed", "cancelled", "expired"]
    requested_at: datetime
    processed_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None

File: apps/landing/services/api/consent.ts
/**
 * Consent and GDPR API service
 */

import { apiClient } from './client'

// Types
export interface CookieConsentRequest {
  visitor_id: string
  necessary: boolean
  functional: boolean
  analytics: boolean
  marketing: boolean
}

export interface CookieConsentResponse {
  id: string
  visitor_id: string
  user_id: string | null
  necessary: boolean
  functional: boolean
  analytics: boolean
  marketing: boolean
  created_at: string
  updated_at: string
}

export interface ConsentStatusResponse {
  terms_accepted: boolean
  terms_accepted_at: string | null
  terms_version: string | null
  privacy_accepted: boolean
  privacy_accepted_at: string | null
  privacy_version: string | null
  marketing_consent: boolean
  marketing_consent_at: string | null
}

export interface UpdateMarketingConsentRequest {
  marketing_consent: boolean
}

export interface UpdateMarketingConsentResponse {
  marketing_consent: boolean
  marketing_consent_at: string
  message: string
}

export interface ConsentHistoryItem {
  id: string
  consent_type: string
  consent_given: boolean
  version: string | null
  created_at: string
}

export interface ConsentHistoryResponse {
  history: ConsentHistoryItem[]
  total: number
}

export interface GdprExportRequestResponse {
  id: string
  status: string
  requested_at: string
  expires_at: string | null
  message: string
}

export interface GdprRequestStatusResponse {
  id: string
  request_type: 'export' | 'deletion'
  status: 'pending' | 'processing' | 'completed' | 'cancelled' | 'expired'
  requested_at: string
  processed_at: string | null
  expires_at: string | null
}

export interface GdprDeletionRequestResponse {
  id: string
  status: string
  requested_at: string
  scheduled_deletion_at: string
  message: string
}

export interface GdprDeletionCancelResponse {
  message: string
  cancelled: boolean
}

export const consentService = {
  // Cookie Consent
  saveCookieConsent: async (data: CookieConsentRequest): Promise<CookieConsentResponse> => {
    return apiClient.post<CookieConsentResponse>('/consent/cookies', data)
  },

  getCookieConsent: async (visitorId: string): Promise<CookieConsentResponse | null> => {
    try {
      return await apiClient.get<CookieConsentResponse>(`/consent/cookies?visitor_id=${encodeURIComponent(visitorId)}`)
    } catch {
      return null
    }
  },

  // User Consent Status
  getConsentStatus: async (): Promise<ConsentStatusResponse> => {
    return apiClient.get<ConsentStatusResponse>('/consent/me')
  },

  updateMarketingConsent: async (data: UpdateMarketingConsentRequest): Promise<UpdateMarketingConsentResponse> => {
    return apiClient.put<UpdateMarketingConsentResponse>('/consent/me', data)
  },

  getConsentHistory: async (limit = 50, offset = 0): Promise<ConsentHistoryResponse> => {
    return apiClient.get<ConsentHistoryResponse>(`/consent/history?limit=${limit}&offset=${offset}`)
  },

  // GDPR Data Export
  requestDataExport: async (): Promise<GdprExportRequestResponse> => {
    return apiClient.post<GdprExportRequestResponse>('/gdpr/export-request')
  },

  getExportStatus: async (): Promise<GdprRequestStatusResponse> => {
    return apiClient.get<GdprRequestStatusResponse>('/gdpr/export-status')
  },

  downloadExport: async (token: string): Promise<Blob> => {
    const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/gdpr/export-download?token=${encodeURIComponent(token)}`, {
      headers: {
        Authorization: `Bearer ${localStorage.getItem('access_token')}`,
      },
    })
    if (!response.ok) {
      throw new Error('Failed to download export')
    }
    return response.blob()
  },

  // GDPR Account Deletion
  requestAccountDeletion: async (): Promise<GdprDeletionRequestResponse> => {
    return apiClient.post<GdprDeletionRequestResponse>('/gdpr/delete-request')
  },

  cancelAccountDeletion: async (): Promise<GdprDeletionCancelResponse> => {
    return apiClient.post<GdprDeletionCancelResponse>('/gdpr/delete-cancel')
  },

  getDeletionStatus: async (): Promise<GdprRequestStatusResponse> => {
    return apiClient.get<GdprRequestStatusResponse>('/gdpr/delete-status')
  },
}
